File: examples/sdk/python/observations.py
#!/usr/bin/env python3
"""Fetch observation history for a task using the Python SDK."""

import os
import sys

from soulbrowser_sdk.client import SoulBrowserClient

BASE_URL = os.environ.get('SOULBROWSER_API_BASE', 'http://127.0.0.1:8801')


def resolve_task_id(client: SoulBrowserClient) -> str:
    if len(sys.argv) > 1:
        return sys.argv[1]
    tasks = list(client.list_tasks(limit=5))
    if not tasks:
        raise RuntimeError('No tasks available; pass a task id as the first argument.')
    for task in tasks:
        status = (task.get('last_status') or '').lower()
        if status == 'running':
            return task['task_id']
    return tasks[0]['task_id']


def main() -> None:
    limit = int(sys.argv[2]) if len(sys.argv) > 2 else 20
    with SoulBrowserClient(base_url=BASE_URL) as client:
        task_id = resolve_task_id(client)
        print(f'[observations] task={task_id} limit={limit}')
        response = client.get_task_observations(task_id, limit=limit)
        observations = response.get('observations', [])
        if not observations:
            print('No observations found.')
            return
        for idx, item in enumerate(observations, 1):
            obs_type = item.get('observation_type') or item.get('content_type') or 'unknown'
            recorded = item.get('recorded_at') or 'n/a'
            dispatch = item.get('dispatch_label') or '-'
            step = item.get('step_id') or '-'
            print(f"#{idx} [{obs_type}] step={step} dispatch={dispatch} recorded={recorded}")
            if item.get('bbox') is not None:
                print('    bbox:', item['bbox'])
            if item.get('screenshot_path'):
                print('    screenshot:', item['screenshot_path'])


if __name__ == '__main__':
    try:
        main()
    except Exception as exc:  # pragma: no cover - example script
        print(f'Error: {exc}', file=sys.stderr)
        sys.exit(1)

File: examples/sdk/typescript/observations.mjs
#!/usr/bin/env node
import WebSocket from 'ws';
import { SoulBrowserClient } from '../../sdk/typescript/dist/index.js';

const baseUrl = process.env.SOULBROWSER_API_BASE ?? 'http://127.0.0.1:8801';
const taskIdArg = process.argv[2];
const limit = Number(process.argv[3] ?? '20');

async function resolveTaskId(client) {
  if (taskIdArg) {
    return taskIdArg;
  }
  const tasks = await client.listTasks(5);
  const active = tasks.find((task) => task.last_status?.toLowerCase() === 'running');
  if (active) {
    return active.task_id;
  }
  if (tasks.length > 0) {
    return tasks[0].task_id;
  }
  throw new Error('No tasks available. Provide a task id as the first argument.');
}

async function run() {
  const client = new SoulBrowserClient({ baseUrl, WebSocketClass: WebSocket });
  const taskId = await resolveTaskId(client);
  console.log(`[observations] querying task ${taskId} (limit=${limit})`);
  const response = await client.getTaskObservations(taskId, limit);
  if (!response.success) {
    console.error('API returned failure payload:', response);
    process.exit(1);
  }
  if (!response.observations.length) {
    console.log('No observations found.');
    return;
  }
  response.observations.forEach((item, index) => {
    const type = item.observation_type || item.content_type || 'unknown';
    const recorded = item.recorded_at || 'n/a';
    const desc = item.dispatch_label || item.label || item.step_id || 'entry';
    console.log(`#${index + 1} [${type}] step=${item.step_id ?? '-'} dispatch=${item.dispatch_label ?? '-'} recorded=${recorded}`);
    if (item.bbox) {
      console.log('    bbox:', JSON.stringify(item.bbox));
    }
    if (item.screenshot_path) {
      console.log('    screenshot:', item.screenshot_path);
    }
  });
}

run().catch((err) => {
  console.error(err);
  process.exit(1);
});
